// @four-leaf-studios/rl-socket-hook/dist/RLProvider.js
import { jsx } from 'react/jsx-runtime';
import { createContext, useRef, useEffect, useMemo, useContext, useSyncExternalStore, useCallback } from 'react';

const RLContext = createContext(null);
const RLProvider = ({ children, url = "ws://localhost:49122", }) => {
    // Strongly-typed storage of known events
    const dataRef = useRef({});
    const subscribersRef = useRef({});
    useEffect(() => {
        const socket = new WebSocket(url);
        socket.onmessage = (e) => {
            try {
                const raw = JSON.parse(e.data);
                const entries = Array.isArray(raw)
                    ? raw
                    : raw.event && raw.data
                        ? [[raw.event, raw.data]]
                        : Object.entries(raw);
                for (const [eventName, payload] of entries) {
                    // store payload
                    dataRef.current[eventName] = payload;
                    // notify subscribers
                    (subscribersRef.current[eventName] ?? []).forEach((cb) => cb());
                }
            }
            catch {
                console.error("Invalid WS data");
            }
        };
        return () => {
            socket.close();
        };
    }, [url]);
    // Expose stable store API
    const store = useMemo(() => ({
        getSnapshot: (eventName) => dataRef.current[eventName],
        subscribe: (eventName, callback) => {
            const subs = (subscribersRef.current[eventName] ??= []);
            subs.push(callback);
            return () => {
                subscribersRef.current[eventName] = subs.filter((c) => c !== callback);
            };
        },
    }), []);
    return jsx(RLContext.Provider, { value: store, children: children });
};
/**
 * Subscribe to the full event payload. Returns `PayloadStorage[E] | undefined`.
 */
function useEvent(eventName) {
    const store = useContext(RLContext);
    if (!store)
        throw new Error("useEvent must be inside <RLProvider>");
    // 2) (optional) explicitly tell TS what snapshot type is
    return useSyncExternalStore((cb) => store.subscribe(eventName, cb), () => store.getSnapshot(eventName));
}
/**
 * Tiny deep-equal for JavaScript values
 */
function deepEqual(a, b) {
    if (Object.is(a, b))
        return true;
    if (typeof a !== "object" ||
        a === null ||
        typeof b !== "object" ||
        b === null) {
        return false;
    }
    const aKeys = Object.keys(a);
    const bKeys = Object.keys(b);
    if (aKeys.length !== bKeys.length)
        return false;
    for (const key of aKeys) {
        if (!deepEqual(a[key], b[key]))
            return false;
    }
    return true;
}
/**
 * Subscribe to a selected slice of the event payload. Only re-renders when that slice changes.
 * Uses a built-in deep-equality check by default (no lodash).
 */
function useEventSelector(
/** Event key to subscribe to */
eventName, 
/** Selector that picks a slice of the payload */
selector, 
/** Equality function to compare selector results; defaults to deepEqual */
isEqual = deepEqual) {
    const store = useContext(RLContext);
    const lastRef = useRef({});
    const subscribe = useCallback((cb) => store.subscribe(eventName, cb), [eventName, store]);
    const getSnapshot = useCallback(() => {
        const raw = store.getSnapshot(eventName);
        const next = selector(raw);
        // deep-equal by default
        if (lastRef.current.val !== undefined &&
            isEqual(next, lastRef.current.val)) {
            return lastRef.current.val;
        }
        lastRef.current.val = next;
        return next;
    }, [eventName, selector, store, isEqual]);
    return useSyncExternalStore(subscribe, getSnapshot);
}

export { RLContext, RLProvider, useEvent, useEventSelector };
